lsusb.c

```diff
@@ -2200,29 +2200,34 @@ static void dump_ccid_device(const unsigned char *buf)
 		fputs("          Auto clock change\n", stdout);
 	if ((us & 0x0020))
 		fputs("          Auto baud rate change\n", stdout);
-	if ((us & 0x0040))
+	if ((us & (0x0040 | 0x0080)) == 0x0040)
 		fputs("          Auto parameter negotiation made by CCID\n", stdout);
-	else if ((us & 0x0080))
+	else if ((us & (0x0040 | 0x0080)) == 0x0080)
 		fputs("          Auto PPS made by CCID\n", stdout);
 	else if ((us & (0x0040 | 0x0080)))
 		fputs("        WARNING: conflicting negotiation features\n", stdout);
 
 	if ((us & 0x0100))
 		fputs("          CCID can set ICC in clock stop mode\n", stdout);
 	if ((us & 0x0200))
-		fputs("          NAD value other than 0x00 accepted\n", stdout);
+		fputs("          NAD value other than 0x00 accepted (T=1)\n", stdout);
 	if ((us & 0x0400))
-		fputs("          Auto IFSD exchange\n", stdout);
+		fputs("          Auto IFSD exchange (T=1)\n", stdout);
 
-	if ((us & 0x00010000))
+	if ((us & 0x00070000) == 0)
+		fputs("          Character level exchange\n", stdout);
+	else if ((us & 0x00070000) == 0x00010000)
 		fputs("          TPDU level exchange\n", stdout);
-	else if ((us & 0x00020000))
+	else if ((us & 0x00070000) == 0x00020000)
 		fputs("          Short APDU level exchange\n", stdout);
-	else if ((us & 0x00040000))
+	else if ((us & 0x00070000) == 0x00040000)
```

Comment:
again, look at the next if statment, I think it's not correct here anymore, or something needs to be done.

Reply:
Again similar as above, this construction checks that only at most one of the bits in the mask 0x70000 result is set. Multiple bits is a violation of the USB-CCID spec.

Reply:
Agreed